src/langsmith/custom-encryption.mdx

```diff
@@ -0,0 +1,407 @@
+    tenant_id = ctx.metadata.get("tenant_id", "default")
+    key_id = ctx.metadata.get("key_id")
+
+    encrypted_data = {}
+    for key, value in data.items():
+        if key in SENSITIVE_FIELDS or key.startswith(SENSITIVE_PREFIX):
+            # Encrypt PII (messages, emails, phone numbers, etc.)
+            response = kms_client.encrypt(
+                KeyId=key_id,
+                Plaintext=str(value).encode(),
+                EncryptionContext={'tenant_id': tenant_id}
+            )
+            encrypted_data[key] = base64.b64encode(
+                response['CiphertextBlob']
+            ).decode()
+        else:
+            # Keep fields like "owner" or "status" unencrypted for search/filtering
+            encrypted_data[key] = value
+
+    return encrypted_data
+
+
+@encryption.decrypt.json
+async def decrypt_json_data(ctx: EncryptionContext, data: dict) -> dict:
+    """Decrypt PII fields."""
+    tenant_id = ctx.metadata.get("tenant_id", "default")
+
+    decrypted_data = {}
+    for key, value in data.items():
+        if (key in SENSITIVE_FIELDS or key.startswith(SENSITIVE_PREFIX)) and isinstance(value, str):
+            ciphertext = base64.b64decode(value.encode())
+            response = kms_client.decrypt(
+                CiphertextBlob=ciphertext,
+                EncryptionContext={'tenant_id': tenant_id}
+            )
+            decrypted_data[key] = response['Plaintext'].decode()
+        else:
+            decrypted_data[key] = value
+
+    return decrypted_data
+```
+
+<Note>
+When data is encrypted, the `ctx.metadata` (from the `X-Encryption-Context` header) is stored alongside it. On decryption, this context is automatically restored—callers don't need to pass the header again when reading data.
+</Note>
+
+### 2. Configure in langgraph.json
+
+Add the path to your encryption module in your `langgraph.json`:
+
+```json {7-9}
+{
+  "dependencies": ["."],
+  "graphs": {
+    "agent": "./agent.py:graph"
+  },
+  "env": ".env",
+  "encryption": {
+    "path": "./encryption.py:encryption"
+  }
+}
+```
+
+### 3. Pass encryption context in requests
+
+When making requests to your deployment, include the `X-Encryption-Context` header with a base64-encoded JSON object containing your encryption parameters:
+
+<Tabs>
+  <Tab title="Python Client">
+```python
+import base64
+import json
+from langgraph_sdk import get_client
+
+# Define encryption context
+encryption_context = {
+    "tenant_id": "customer-123",
+    "key_id": "arn:aws:kms:us-east-1:123456789:key/abc-def-123"
+}
+
+# Encode as base64
+encoded_context = base64.b64encode(
+    json.dumps(encryption_context).encode()
+).decode()
+
+# Create client with encryption context header
+client = get_client(
+    url="http://localhost:2024",
+    headers={"X-Encryption-Context": encoded_context}
+)
+
+# Create a thread with encrypted metadata
+thread = await client.threads.create(
+    metadata={
+        "owner": "user-456",  # Unencrypted (for search)
```

Comment:
I think `owner` would actually typically be provided via auth middleware so it isn't spoofed by a client fwiw